### PR TITLE
New UCL API Decorator

### PR DESCRIPTION
**Pull request by ChristopherHammond13**
## What does this PR do?
- Unifies rate limiting to work on OAuth, standard tokens and temporary tokens. OAuth previously was not limited.
- Allows non-personal data to be obtained with an OAuth token to eliminate needing two token types
- Cleans up a bunch of code for cleanliness.
- Creation of a new `common` app to centralise some of our common code.

## ✅ Pull Request checklist

- [x] Is this code complete?
- [x] Are tests done/modified?

## 📃 Link to PR updating documentation (if applicable)
N/A for now, but there will be new docs when the rate limiting is properly in place.

## 🚨 Is this a breaking change for API clients?
Yes, as I have renamed the rate limit header to more similarly follow what GitHub do: https://developer.github.com/v3/#rate-limiting

## :squirrel: Deploy notes
Just a `git pull`, `pip install -r requirements.txt` and `supervisorctl restart uclapi`.

## Anything else
This should be ready to go after a code review.

### CODE REVIEW

File: backend/uclapi/common/decorators.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+            "ok": False,
+            "error":
+                "The token is inactive as the user has revoked "
+                "your app's access to their data."
+        })
+        response.status_code = 400
+        return response
+
+    scopes = Scopes()
+    for s in required_scopes:
```
[the remark this answers is not in the document]
Comment (ChristopherHammond13) [reply backend/uclapi/common/decorators.py]: This has been sorted now. Come on, Hound. Don't do this to me :(

File: backend/uclapi/roombookings/tests.py [hunk 2]
```diff
@@ -178,39 +181,36 @@ class DoesTokenExistTestCase(TestCase):
     def setUp(self):
         mock = unittest.mock.Mock()
         mock.status_code = 200
-        self.dec_view = does_token_exist(
-            unittest.mock.Mock(return_value=mock)
-        )
+
```
[the remark this answers is not in the document]
Comment (ChristopherHammond13) [reply backend/uclapi/roombookings/tests.py]: This code doesn't event exist any more

File: backend/uclapi/common/decorators.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChristopherHammond13) [reply backend/uclapi/common/decorators.py]: again, nothing has changed kiddo

File: backend/uclapi/common/decorators.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
+import datetime
+from functools import wraps
```
Comment (jermenkoo) [inline backend/uclapi/common/decorators.py]: This import should be after `import redis`.

File: backend/uclapi/common/decorators.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+    headers = request.META
+    version_headers = {}
+    regex = re.compile("^HTTP_UCLAPI_.*_VERSION$")
+    for header in headers:
+        if regex.match(header):
+            version_headers[header] = headers[header]
+
+    queryparams = dict(request.GET)
+
+    if token_type == "oauth":
```
Comment (jermenkoo) [inline backend/uclapi/common/decorators.py]: The `parameters` dict for `general` and `oauth` are the same, can't they be merged?

File: backend/uclapi/common/decorators.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+            "error": "Token does not exist."
+        })
+        response.status_code = 400
+        return response
+
+    # No issues, so return the token
+    return token
+
+
+def uclapi_protected_endpoint(personal_data=False, required_scopes=[]):
```
Comment (jermenkoo) [inline backend/uclapi/common/decorators.py]: I would be wary using mutable variables as arguments. (and linter should be complaining too)

File: backend/uclapi/common/helpers.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class PrettyJsonResponse(JsonResponse):
+    def __init__(self, data, rate_limiting_data=None):
+        # Calls JsonResponse's constructure and requests 4 line indenting
+        super().__init__(data, json_dumps_params={'indent': 4})
+
+        # Adds rate limiting headers from a passed view kwargs
+        if rate_limiting_data:
+            if 'X-RateLimit-Limit' in rate_limiting_data:
```
Comment (jermenkoo) [inline backend/uclapi/common/helpers.py]: Maybe we could extract those to a set / similar and check for them in a loop?

File: backend/uclapi/common/helpers.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+from binascii import hexlify
+
+from django.http import JsonResponse
+
+import os
```
Comment (jermenkoo) [inline backend/uclapi/common/helpers.py]: These 2 imports should be above `from binascii ...`.

File: backend/uclapi/common/decorators.py (hunk 5 above)
Comment (ChristopherHammond13) [reply backend/uclapi/common/decorators.py]: What's the solution to this in this case? Remember that `required_scopes` is never written to by the decorator, only read from.

**Comment by ChristopherHammond13** [comment]
Testing Jenkins access, ignore!

**Comment by ChristopherHammond13** [comment]
Can one of the admins verify this patch?

File: backend/uclapi/common/decorators.py [hunk 8]
```diff
@@ -55,7 +56,7 @@ def log_api_call(request, token, token_type):
 
     queryparams = dict(request.GET)
 
-    if token_type == "oauth":
+    if token_type in("general", "oauth"):
```
Comment (jermenkoo) [inline backend/uclapi/common/decorators.py]: `if token_type in {"general", "oauth"}:`

File: backend/uclapi/common/helpers.py [hunk 9]
```diff
@@ -1,24 +1,26 @@
...
-import textwrap
-
 
 class PrettyJsonResponse(JsonResponse):
     def __init__(self, data, rate_limiting_data=None):
         # Calls JsonResponse's constructure and requests 4 line indenting
         super().__init__(data, json_dumps_params={'indent': 4})
 
         # Adds rate limiting headers from a passed view kwargs
+        rate_limit_headers = set([
```
Comment (jermenkoo) [inline backend/uclapi/common/helpers.py]: Construct the set inline like `{'a', 'b', ...}`

**Comment by ChristopherHammond13** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.uclapi.com/job/UCLAPI/12/
Jenkins statics successfully built.

**Comment by ChristopherHammond13** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.uclapi.com/job/UCLAPI/13/
Jenkins statics successfully built.

**Comment by ChristopherHammond13** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.uclapi.com/job/UCLAPI/14/
Jenkins statics successfully built.

**Comment by ChristopherHammond13** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.uclapi.com/job/UCLAPI/15/
Jenkins statics successfully built.

**Comment by ChristopherHammond13** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.uclapi.com/job/UCLAPI/16/
Jenkins statics successfully built.

**Pull request merged**